sort chan_no lists before comparing equality

joshbear opened the pull request:
Thanks for contributing a pull request! Please make sure you have read the
[contribution guidelines](https://martinos.org/mne/stable/contributing.html)
before submitting.

Please be aware that we are a loose team of volunteers so patience is
necessary. Assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Again, thanks for contributing!

#### Reference issue
Fixes #6032


#### What does this implement/fix?
sorts the channel numbers from the config and pdf files prior to comparing.


#### Additional information
n/a

larsoner commented on mne/io/bti/bti.py:
```diff
@@ -914,8 +914,8 @@ def _read_bti_header(pdf_fname, config_fname, sort_by_ch_name=True):
-        match = [c['chan_no'] for c in chans_cfg] == \
-                [c['chan_no'] for c in chans]
+        match = sorted([c['chan_no'] for c in chans_cfg]) == \
+                sorted([c['chan_no'] for c in chans])
```
I suspect that this will not be enough, and that you actually need to reorder one to match the other. See a few lines below, we just `zip` over them -- this implies they are equivalently ordered already.

joshbear replied:
Ah yes, I see that. That will be a problem. Do you think unmatched orders will be a problem elsewhere as well? I can work something to reorder these, although it probably won't be done as elegantly as the code deserves.

joshbear replied:
Interestingly, it looks like there is a typo on line 718 as well. low_frew instead of low_freq. Ran into this problem after sorting when it got to line 1177 and is examining the process steps. Making a note here partially so that I don't forget to note it while I'm working on this.

joshbear commented:
Changed to sort by chan_no values. This should allow the subsequent zip to function properly. I tested this locally and it appears to function as expected for my data. Fixed the typo.

larsoner commented on mne/io/bti/bti.py:
```diff
@@ -913,6 +913,10 @@ def _read_bti_header(pdf_fname, config_fname, sort_by_ch_name=True):
 
+        # sort chans_cfg and chans
+        chans = sorted(chans, key=lambda k: k['chan_no'])
+        chans_cfg = sorted(chans_cfg, key=lambda k: k['chan_no'])
```
If the `chan_no` always corresponds to the index in the data, then this is the right thing to do.

But it's possible a different scheme is used, where `chans_cfg` or `chans` is already ordered to match the data, and `chan_no` does not necessarily give the index to the data (e.g., `chans[0]` corresponds to the first "row" of stored data, but `chans[0]['chan_no']` could be some other number like 2), then it's not. Any ideas from the format?

@dengemann I think you originally wrote the BTi code, any idea about this, or how to check?

joshbear replied:
Since the subject I ran across this error on was the first to throw an exception, I suspect (but have not confirmed) that the order of 'chan_no' does much the row indices in other subjects. In this one, you are correct that the initial and 'chan_no' do not correlate. For example, in my data, 'chans' is derived as done by MNE while 'nchans' is the sorting result:

>>> chans[0]
{'chan_label': 'A22', 'chan_no': 8, 'attributes': 0, 'scale': 1.0, 'yaxis_label': 'T', 'valid_min_max': 1, 'ymin': -4.868604127669051e-12, 'ymax': 4.3606116638294954e-12, 'index': 0, 'checksum': -2876, 'off_flag': 'OFF+', 'offset': 0.0}

>>> nchans[0]
{'chan_label': 'TRIGGER', 'chan_no': 1, 'attributes': 2, 'scale': 1.0, 'yaxis_label': 'bit', 'valid_min_max': 1, 'ymin': 0.0, 'ymax': 8192.0, 'index': 303, 'checksum': -1615, 'off_flag': 'OFF-', 'offset': 0.0}

In the original 'chans', the TRIGGER (chan_no=1) channel is at index 303:
>>> chans[303]
{'chan_label': 'TRIGGER', 'chan_no': 1, 'attributes': 2, 'scale': 1.0, 'yaxis_label': 'bit', 'valid_min_max': 1, 'ymin': 0.0, 'ymax': 8192.0, 'index': 303, 'checksum': -1615, 'off_flag': 'OFF-', 'offset': 0.0}

The sorting maintains the integrity of full dictionary, so as long as everything is referenced by 'chan_no' and not by the index, it would be fine. If however the index is used without comparing chan_no, it could introduce a mismatch...

ebeich replied:
Sorting the chans and chans_cfg using chan_no should be fine. I used chans.sort(key=lambda x: x['chan_no'])
and chans_cfg.sort(key=lambda x: x['chan_no']), which is probably the same as sorted().
For the background: the config-file contains all channels showing up in a run, a data file (pdf) usually contains just a subset of channels. The value for a channel (chan_no) in a time slice is given by its index (stored in the pdf)
and the (chan_no, chan_index) relation must be maintained, but the sequence of the channel structures
(dftk_channel in the config and dftk_channel_ref in the pdf) is arbitrary. It just happens that during acquisition
channels are stored in the same sequence as in the config file.

larsoner replied:
Great, thanks for the background @ebeich and the +1 (as I understand it) for this change

larsoner commented:
@joshbear can you add an update to `whats_rew.rst` in the `BUG` section mentioning this fix?

larsoner commented:
Closing for #6300
